### DISCUSSION TITLE
MMM.optimal_allocation_dict not giving correct values

### DISCUSSION THREAD

**Discussion by emil-matti117** [pymc-labs/pymc-marketing#1234] [Q&A]
Hey everyone! I'm using the budget allocation function `MMM.allocate_budget_to_maximize_response` and when trying to print out the `mmm.optimal_allocation_dict` , it is giving me an equal budget across all of my channels.
My initial thought is because I did not assign budget bounds to any channel this was the reason, but based on this [example](https://www.pymc-marketing.io/en/stable/notebooks/mmm/mmm_budget_allocation_example.html#example-use-cases) in the documentation, that shouldn't be the case. Any thoughts here?

![image](https://github.com/user-attachments/assets/b4f0dfea-ea84-4f5f-8396-48849de842a6)
![image](https://github.com/user-attachments/assets/dd8bcf66-4b87-4570-a3e4-b9f134f8ff44)

For reference, the Allocation column is the `optimal_allocation_dict` but it is not matching the blue bars in the `mmm.plot_budget_allocation` function (which the source code points to `optimal_allocation_dict`.

**Comment by williambdean** [comment]
@cetagostini would know best

**Comment by cetagostini** [comment]
Hello @emil-matti117 

Can you share the code you are using? A few lines would be enough. I want to see how did you get the dataset and calculate the response.

**Reply by emil-matti117** [reply]
Hey @cetagostini ! Thanks for this, here's the code:

```
model_load = MMM.load(model_path)

channels = model_load.channel_columns

response = model_load.allocate_budget_to_maximize_response(
            budget=27_000_000, 
            time_granularity='monthly',
            num_periods=1,
            budget_bounds=None
        )

print(model_load.optimal_allocation_dict)
```


{'AFFILIATE': 2112633.5375013724,
 'DISCOVERY': 2112633.5392965553,
 'DISPLAY': 2112633.5375013724,
 'Performance_Email': 2112633.5375005724,
 'PMAX': 2112633.5375013724,
 'RADIO': 2112633.5427716007,
 'SEM_BRAND': 2112633.5374436025,
 'SEM_NONBRAND': 2112633.53838007,
 'SOCIAL_Lower': 2112633.542714631,
 'SOCIAL_Middle': 2112633.5389476637,
 'SOCIAL_Upper': 2112633.5364613836,
 'TV': 2112633.5374205266,
 'YOUTUBE': 2112633.5365592754}

**Reply by cetagostini** [reply]
Hey @emil-matti117 , 

I think maybe the object `model_load.optimal_allocation_dict` is not being overwritten that's why you see different results in the dictionary and the plot, it's rare, but it could be the reason. I'd have to look into it in depth to be able to answer.

Regarding why the optimizer doesn't move it's possible it thinks it's in the minimum because the result of computing the gradient is quite small. You could try to verify this, modifying the source code.

[We are allowing a more flexible optimizer in the next release](https://github.com/pymc-labs/pymc-marketing/pull/1151), with more options to debug and work around this types of issues.

Meantime, you can play with minimize parameters:

```python
minimize_kwargs = {
        "method": "SLSQP",
        "options": {
            "ftol": 1e4,
            "eps": 1e3,
            "maxiter": 3_000,
        },
    }
```

Basically depending on the data scale you are working on, you could decrease the tolerance or the increase the size steps on how the budget varies to force the optimizer to move within a certain degree of flexibility based on your data. Allowing it to get out of the initial guess.

More information [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize)!

**Reply by emil-matti117** [reply]
Hey @cetagostini - thanks for this! I tried passing in the `minimize_kwargs` argument but it is giving me a 
`TypeError: MMM.allocate_budget_to_maximize_response() got an unexpected keyword argument 'options'`

```
minimize_kwargs = {
        "method": "SLSQP",
        "options": {
            "ftol": 1e4,
            "eps": 1e3,
            "maxiter": 3_000,
        }}
response = model_load.allocate_budget_to_maximize_response(
            budget=27_464_236, 
            time_granularity='monthly',
            num_periods=1,
            budget_bounds=None, 
            **minimize_kwargs
        )
```

any thoughts here?

**Reply by AlfredoJF** [reply]
@emil-matti117 I think the `minimize_kwargs` settings in `allocate_budget_to_maximize_response` is the issue. We should unpack the `minimize_kwargs` dict until it's called by the Scipy `minimize` in the `allocate_budget` method but @cetagostini might have a better answer than me

**Reply by emil-matti117** [reply]
Thanks @AlfredoJF - tried the `allocate_budget_to_maximize_response` on a different model and the `optimal_allocation_dict` is still coming back equal. 

I replicated the model provided in the [documentation](https://www.pymc-marketing.io/en/stable/notebooks/mmm/mmm_example.html) and ran an allocation on it and the `optimal_allocation_dict` is returning correctly. Unsure where the issue lies.

**Reply by tim-mcwilliams** [reply]
Working alongside @emil-matti117 here, we have tried to change up the parameters when calling the `allocate_budget_to_maximize_response` function. When changing the `time_granularity`, `num_periods`, and also the `noise_level` the `optimal_allocation_dict` returns the same results. I would figure if we change the allocation parameters that we would at least get different results. 

Noting that the `plot_budget_allocation` does return different results when changing those parameters.

**Reply by cetagostini** [reply]
Hey @AlfredoJF thats totally correct.

@emil-matti117 what parameters have you been using? Can you share? Whats the budget you are using? Are you able to access to minimize object? Observe the results there should help to understand whats actually happening during the optimization.

@tim-mcwilliams did you tried increasing iterations as well?

**Reply by tim-mcwilliams** [reply]
@cetagostini, I have not as of yet. Where would we increase the iterations at? Not seeing an arg in the `allocate_budget_to_maximize_response` for that. If that has to be done in the `**minimize_kwargs`, right now that isn't working for us. It's yelling at us about a `TypeError`.

**Reply by emil-matti117** [reply]
```
response1 = model_load.allocate_budget_to_maximize_response(
            budget=27_464_236, 
            time_granularity='monthly',
            num_periods=1,
            budget_bounds=None
        )

response2 = model_load.allocate_budget_to_maximize_response(
            budget=20_000_000, 
            time_granularity='monthly',
            num_periods=3,
            budget_bounds=None,
            noise_level =.02
        )

response3 = model_load.allocate_budget_to_maximize_response(
            budget=27_464_236, 
            time_granularity='quarterly',
            num_periods=3,
            budget_bounds=budget_bounds 
        )
```

@cetagostini these are examples of iterations we've been using. We are also unsure how to access the `minimize_kwargs` object to pass that to the function, do you mind showing us how we'd go about that as well?

Thanks!

**Reply by cetagostini** [reply]
Are you changing granularity for the same model? Even when model was train with the same granularity? @emil-matti117 

On the other hand, are you trying something like the following? @tim-mcwilliams 
```python
results = mmm.allocate_budget_to_maximize_response(
    budget=100000,
    time_granularity="weekly",
    num_periods=4,
    budget_bounds=budget_bounds,
    minimize_kwargs={
        "method": "SLSQP",
        "options": {
            "maxiter": 3000,
            "disp": True
        }
    }
)
```

**Reply by tim-mcwilliams** [reply]
@cetagostini, Yea we are changing the granularity for the same model. To make sure I am fully understanding here, if the model was trained on weekly data can we only use the weekly time granularity for the allocator??

 Yea, we tried that exact block and are getting the following error - `TypeError: MMM.allocate_budget_to_maximize_response() got an unexpected keyword argument 'options'`

**Comment by emil-matti117** [comment]
@cetagostini yes, we've tried every time granularity and am still getting the same results. Also, the above minimize_kwargs argument is giving me the following error:

`TypeError: MMM.allocate_budget_to_maximize_response() got an unexpected keyword argument 'minimize_kwargs'`
